osbuild/sources.py

```diff
@@ -3,12 +3,51 @@
 import os
 import json
 import tempfile
+from typing import List, Dict
 
+from osbuild.formats_common import OSbuildError
 from . import host
 from .objectstore import ObjectStore
 from .util.types import PathLike
 
 
+class SourceResult():
+
+    def __init__(self, source=None):
+        self.source = source
+        self.error = None
+
+    def to_dict(self) -> Dict:
+        return {
+            "source": self.source,
+            "error": self.get_error_message()
+        }
+
+    def get_error_message(self):
+        return self.error
+
+    @staticmethod
+    def from_dict(obj: Dict):
+        ret = SourceResult()
+        ret.source = obj["source"]
+        ret.error = obj.get("error", None)
+        return ret
+
+    def explode_if_error(self):
+        if self.error:
+            raise SourceException(self)
+
+
+class SourceException(OSbuildError):
```

Comment:
We probably want to define some well known conditions here. Which ones is obviously tricky and we can just start with the generic one (`FAILED` or something), and then start to add ones that are actionable or we care about?
In proper OOP we would obviously do this via inheritance `SourceTimeoutError`, etcpp, but then we cannot handle multiple items and per-item specific errors.

Reply:
I chose a not `OOP` variant of that.
Every possible error is centralized in sources.py. `SourceErrorKind` is an `enum` that contains the list of possible errors. When an error occurs in a subcommand, it instanciate a `SourceError` object with as a field of the constructor a `SourceErrorKind` that defined what happened.
The sourceError is serialized via its` to_dict() `method (from the SourceService class) and is reconstructed and raised in the `raise_error()` one (from the Source class).

In the main, when a SourceError is caught, the format code invokes the format method from OSBuildError and produces the following json

```
        return {
            "type": "error",
            "error": {
                    "id": self.ID,
                    "code": self.code,
                    "reason": str(self),
                    "details": self.details
            }
        }
```

The source's error kind is then used to generate the precise ID of the error while it's name in the `enum` serves to complete the `code` field of the json. Reason is the human readable version of the error and details contains the rest. (In the case of a SourceError, it is the source)